pygen/pygen_src/isa/riscv_compressed_instr.py

```diff
@@ -0,0 +1,228 @@
+You may obtain a copy of the License at
+http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+"""
+
+import logging
+import vsc
+from pygen_src.isa.riscv_instr import riscv_instr
+from pygen_src.riscv_instr_pkg import (riscv_instr_name_t, riscv_instr_format_t,
+                                       riscv_instr_category_t, riscv_reg_t, imm_t, pkg_ins)
+
+
+@vsc.randobj
+class riscv_compressed_instr(riscv_instr):
+    def __init__(self):
+        super().__init__()
+        self.imm_align = 0
+        self.rs1 = riscv_reg_t.S0
+        self.rs2 = riscv_reg_t.S0
+        self.rd = riscv_reg_t.S0
+        self.is_compressed = 1
+
+    @vsc.constraint
+    def rvc_csr_c(self):
+        # Registers specified by the three-bit rs1’, rs2’, and rd’
+        with vsc.implies(self.format.inside(vsc.rangelist(riscv_instr_format_t.CIW_FORMAT,
+                                                          riscv_instr_format_t.CL_FORMAT,
+                                                          riscv_instr_format_t.CS_FORMAT,
+                                                          riscv_instr_format_t.CB_FORMAT,
+                                                          riscv_instr_format_t.CA_FORMAT))):
+            with vsc.implies(self.has_rs1 == 1):
+                self.rs1.inside(vsc.rangelist(riscv_reg_t.S0, riscv_reg_t.S1, riscv_reg_t.A0,
+                                              riscv_reg_t.A1, riscv_reg_t.A2, riscv_reg_t.A3,
+                                              riscv_reg_t.A4, riscv_reg_t.A5))
+            with vsc.implies(self.has_rs2 == 1):
+                self.rs2.inside(vsc.rangelist(riscv_reg_t.S0, riscv_reg_t.S1, riscv_reg_t.A0,
+                                              riscv_reg_t.A1, riscv_reg_t.A2, riscv_reg_t.A3,
+                                              riscv_reg_t.A4, riscv_reg_t.A5))
+            with vsc.implies(self.has_rd == 1):
+                self.rd.inside(vsc.rangelist(riscv_reg_t.S0, riscv_reg_t.S1, riscv_reg_t.A0,
+                                             riscv_reg_t.A1, riscv_reg_t.A2, riscv_reg_t.A3,
+                                             riscv_reg_t.A4, riscv_reg_t.A5))
+        # _ADDI16SP is only valid when rd == SP
+        with vsc.implies(self.instr_name == riscv_instr_name_t.C_ADDI16SP):
+            self.rd == riscv_reg_t.SP
+        with vsc.implies(self.instr_name.inside(vsc.rangelist(riscv_instr_name_t.C_JR,
+                                                              riscv_instr_name_t.C_JALR))):
+            self.rs1 != riscv_reg_t.ZERO
+            self.rs2 == riscv_reg_t.ZERO
+
+    @vsc.constraint
+    def imm_val_c(self):
+        with vsc.implies(self.imm_type.inside(vsc.rangelist(imm_t.NZIMM, imm_t.NZUIMM))):
+            self.imm[5:0] != 0
```

Comment:
I believe there is a small bug with respect to PyVSC part select: it seems like the part select in PyVSC isn't inclusive in both ends, please look at the below example:

```
import vsc
imm = vsc.rand_bit_t(4)
imm.set_val(2)
print("imm val is: {}\n".format(imm[1:0]))
```
This prints 0 instead of 2 (apparently it only prints `imm[0]`), meaning that all of the partial selections that we have here in the constraints are missing the most significant bit. I'll ask Matt for a quick fix for this.

Reply:
Hi @hodjat91 
Thanks for this.